rust/src/buffer.rs

```diff
@@ -41,24 +42,27 @@ impl<T> Buffer<T> {
         self.len
     }
 
+    /// Get a pointer to the data contained by the buffer
     pub fn data(&self) -> *const T {
         self.data
     }
 
     pub fn slice(&self, start: usize, end: usize) -> &[T] {
         assert!(start <= end);
-        assert!(start <= self.len as usize);
+        assert!(start < self.len as usize);
         assert!(end <= self.len as usize);
         unsafe { slice::from_raw_parts(self.data.offset(start as isize), (end - start) as usize) }
     }
 
     /// Get a reference to the value at the specified offset
     pub fn get(&self, i: usize) -> &T {
+        assert!(i < self.len as usize);
         unsafe { &(*self.data.offset(i as isize)) }
     }
 
-    /// Deprecated method (used by Bitmap)
+    /// Write to a slot in the buffer
     pub fn set(&mut self, i: usize, v: T) {
+        assert!(i < self.len as usize);
         unsafe {
             let p = mem::transmute::<*const T, *mut T>(self.data);
```

Comment:
`transmute` is overkill here - can get away with (safe) `let p = self.data as *mut _`